File: dao_cs/platform/platform.dao/param/StatementParameter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Data;
using platform.dao.type;

namespace platform.dao.param
{
    /// <summary>
    /// 对于In类型的参数，将@variable替换为多个?，然后根据传入参数的多少，
    /// 填入相应数量的？，传向DAS或者DbClient
    /// 其他参数直接替换为？，传向DAS或者DbClient
    /// </summary>
    public sealed class StatementParameter : IParameter
    {

        private DbType dbType = DbType.Boolean;
        private ParameterDirection direction = ParameterDirection.Input;
        private bool isNullable = false;
        private string name;
        private int index;
        private int size;
        private object value;
        private bool isSensitive = false;

        public DbType DbType
        {
            get
            {
                return dbType;
            }
            set
            {
                dbType = value;   
            }
        }

        public ParameterDirection Direction
        {
            get
            {
                return direction;
            }
            set
            {
                direction = value;
            }
        }

        public bool IsNullable
        {
            get
            {
                return isNullable;
            }
            set
            {
                isNullable = value;
            }
        }

        public string Name
        {
            get
            {
                return name;
            }
            set
            {
                name = value;
            }
        }

        public int Index
        {
            get
            {
                return index;
            }
            set
            {
                index = value;
            }
        }

        public int Size
        {
            get
            {
                return size;
            }
            set
            {
                size = value;
            }
        }

        public object Value
        {
            get
            {
                return value;
            }
            set
            {
                this.value = value;
            }
        }

        public bool IsSensitive
        {
            get
            {
                return isSensitive;
            }
            set
            {
                isSensitive = value;
            }
        }


        public param.AvailableType GetFromObject()
        {
            switch (dbType)
            {
                case System.Data.DbType.Boolean:
                    return new param.AvailableType() { current = 0, bool_arg = (bool)value};
                case System.Data.DbType.Byte:
                case System.Data.DbType.SByte:
                case System.Data.DbType.UInt16:
                case System.Data.DbType.Int16:
                case System.Data.DbType.Int32:
                case System.Data.DbType.UInt32:
                case System.Data.DbType.StringFixedLength:
                    return new param.AvailableType(){current = 1, int32_arg = (int)value};
                case System.Data.DbType.Int64:
                case System.Data.DbType.UInt64:
                    return new param.AvailableType() { current = 2, int64_arg = (long)value};
                case System.Data.DbType.DateTime:
                    return new param.AvailableType() { current = 2, int64_arg = ((DateTime)value).Ticks / TimeSpan.TicksPerMillisecond };
                case System.Data.DbType.Single:
                case System.Data.DbType.Double:
                    return new param.AvailableType() { current = 3, double_arg = (double)value };
                case System.Data.DbType.AnsiString:
                case System.Data.DbType.AnsiStringFixedLength:
                case System.Data.DbType.String:
                case System.Data.DbType.Decimal:
                    return new param.AvailableType() { current = 4, string_arg = value.ToString()};
                case System.Data.DbType.Binary:
                    return new param.AvailableType() { current = 5, bytes_arg = (byte[])value};
                case System.Data.DbType.Guid:
                    return new param.AvailableType() { current = 5, bytes_arg = ((Guid)value).ToByteArray() };
                default:
                    return new param.AvailableType() { current = 4, string_arg = value.ToString() };
            }
        }
    }
}

File: das/src/main/java/com/ctrip/sysdev/das/domain/StatementParameter.java
package com.ctrip.sysdev.das.domain;

import java.io.IOException;
import java.math.BigDecimal;
import java.sql.CallableStatement;
import java.sql.Date;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.sql.Time;
import java.sql.Timestamp;

import org.msgpack.packer.Packer;
import org.msgpack.type.Value;
import org.msgpack.unpacker.Unpacker;

import com.ctrip.sysdev.das.domain.enums.DbType;
import com.ctrip.sysdev.das.domain.enums.ParameterDirection;

public class StatementParameter implements Comparable<StatementParameter> {

	private DbType dbType;

	private ParameterDirection direction;

	private boolean nullable;

	private int index;

	private String name;

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	private int size;

	private Value value;

	private boolean sensitive;

	/**
	 * 获取C#传入Db类型
	 * 
	 * @return
	 */
	public DbType getDbType() {
		return dbType;
	}

	/**
	 * 设置C# 对应Db类型
	 * 
	 * @param dbType
	 */
	public void setDbType(DbType dbType) {
		this.dbType = dbType;
	}

	/**
	 * 获取参数方向
	 * 
	 * @return
	 */
	public ParameterDirection getDirection() {
		return direction;
	}

	/**
	 * 设置参数方向
	 * 
	 * @param direction
	 */
	public void setDirection(ParameterDirection direction) {
		this.direction = direction;
	}

	/**
	 * 
	 * @return
	 */
	public boolean isNullable() {
		return nullable;
	}

	/**
	 * 
	 * @param nullable
	 */
	public void setNullable(boolean nullable) {
		this.nullable = nullable;
	}

	public int getIndex() {
		return index;
	}

	public void setIndex(int index) {
		this.index = index;
	}

	/**
	 * 
	 * @return
	 */
	public int getSize() {
		return size;
	}

	/**
	 * 
	 * @param size
	 */
	public void setSize(int size) {
		this.size = size;
	}

	/**
	 * 
	 * @return
	 */
	public Value getValue() {
		return value;
	}

	/**
	 * 
	 * @param value
	 */
	public void setValue(Value value) {
		this.value = value;
	}

	/**
	 * 
	 * @return
	 */
	public boolean isSensitive() {
		return sensitive;
	}

	/**
	 * 
	 * @param sensitive
	 */
	public void setSensitive(boolean sensitive) {
		this.sensitive = sensitive;
	}

	public static StatementParameter createFromUnpack(Unpacker unpacker)
			throws IOException {

		StatementParameter param = new StatementParameter();

		unpacker.readArrayBegin();

		param.setDbType(DbType.fromInt(unpacker.readInt()));

		param.setDirection(ParameterDirection.fromInt(unpacker.readInt()));

		param.setNullable(unpacker.readBoolean());

		param.setSensitive(unpacker.readBoolean());

		param.setIndex(unpacker.readInt());

		param.setSize(unpacker.readInt());

		param.setValue(unpacker.readValue());

		unpacker.readArrayEnd();

		return param;

	}

	public void pack(Packer packer) throws IOException {

//		packer.writeArrayBegin(7);

		packer.write(dbType.getIntVal());

		if (direction == null)
			direction = ParameterDirection.Input;

		packer.write(direction.getIntVal());

		packer.write(nullable);

		packer.write(sensitive);

		packer.write(index);

		packer.write(name);

		packer.write(size);

		packer.write(value);

//		packer.writeArrayEnd();

	}

	public static StatementParameter createFromValue(int index, DbType dbType,
			Value value) {

		StatementParameter param = new StatementParameter();
		param.setIndex(index);
		param.setDbType(dbType);
		param.setValue(value);

		return param;

	}

	public static StatementParameter createFromValue(int index, String name,
			DbType dbType, Value value) {

		StatementParameter param = new StatementParameter();
		param.setIndex(index);
		param.setName(name);
		param.setDbType(dbType);
		param.setValue(value);

		return param;

	}

	@Override
	public int compareTo(StatementParameter o) {
		return this.index - o.getIndex();
	}

	public PreparedStatement setPreparedStatement(PreparedStatement ps)
			throws SQLException {
		
		if(direction == ParameterDirection.Output && ps instanceof CallableStatement){
			((CallableStatement)ps).registerOutParameter(index, DbType.getFromDbType(dbType));
			return ps;
		}

		switch (dbType) {
//		case AnsiString:
//
//			break;
//		case AnsiStringFixedLength:
//			break;
		case Binary:
			ps.setBytes(index, value.asRawValue().getByteArray());
			break;
		case Boolean:
			ps.setBoolean(index, value.asBooleanValue().getBoolean());
			break;
		case Byte:
		case SByte:
			ps.setByte(index, value.asIntegerValue().getByte());
			break;
//		case Currency:
		case Decimal:
			ps.setBigDecimal(index, new BigDecimal(value.asRawValue().getString()));
			break;
//		case Date:
//			ps.setDate(index, new Date(value.asIntegerValue().getLong()));
//			break;
		case DateTime:
//		case DateTime2:
			ps.setTimestamp(index, new Timestamp(value.asIntegerValue()
					.getLong()));
			break;
//		case DateTimeOffset:
//			break;
		case Double:
			ps.setDouble(index, value.asFloatValue().getDouble());
			break;
		case Guid:
			ps.setBytes(index, value.asRawValue().getByteArray());
			break;
		case Int16:
		case UInt16:
			ps.setShort(index, value.asIntegerValue().getShort());
			break;
		case Int32:
		case UInt32:
			ps.setInt(index, value.asIntegerValue().getInt());
			break;
		case Int64:
		case UInt64:
			ps.setLong(index, value.asIntegerValue().getLong());
			break;
//		case Object:
//			break;
		case Single:
			ps.setFloat(index, value.asFloatValue().getFloat());
			break;
		case String:
		case StringFixedLength:
			ps.setString(index, value.asRawValue().getString());
			break;
//		case Time:
//			ps.setTime(index, new Time(value.asIntegerValue().getLong()));
//			break;
//		case VarNumeric:
//			ps.setDouble(index, value.asFloatValue().getDouble());
//			break;
//		case Xml:
//			ps.setString(index, value.asRawValue().getString());
//			break;
		default:
			break;
		}

		return ps;

	}

}
